File: template/golang/dut.go
package UT_{{__TOP_MODULE_NAME__}}


import (
    "xspcomm"
    "container/list"
)

type UT_{{__TOP_MODULE_NAME__}} struct {
    Dut DutUnifiedBase
    Xclock xspcomm.XClock
    Xport xspcomm.XPort
    internalSignals map[string]xspcomm.XData
    internalSignalsList map[string]*list.List
    Xcfg xspcomm.XSignalCFG
    // Pins
{{__XDATA_DECL__}}
    // SubPorts
{{__XPORT_CASCADED_DEC__}}
}


func NewUT_{{__TOP_MODULE_NAME__}}(a ...interface{}) *UT_{{__TOP_MODULE_NAME__}}{
    self := &UT_{{__TOP_MODULE_NAME__}}{}
    self.internalSignals = make(map[string]xspcomm.XData)
	if len(a) == 1 {
        strings, ok := a[0].([]string)
        if ok {
            args := NewStringVector()
            for _, v := range strings {
                args.Add(v)
            }
            self.Dut = NewDutUnifiedBase(args)
        }else{
            self.Dut = NewDutUnifiedBase(a ...)
        }
    }else{
        self.Dut = NewDutUnifiedBase(a ...)
    }
    self.Xclock = xspcomm.NewXClock(self.Dut.GetPxcStep(), self.Dut.GetPSelf())
    self.Xport = xspcomm.NewXPort()
    self.Xcfg = xspcomm.NewXSignalCFG(self.Dut.GetXSignalCFGPath(), self.Dut.GetXSignalCFGBasePtr())
    // Create
{{__XDATA_INIT__}}
    // Bind
{{__XDATA_BIND__}}
    // Add to port
{{__XPORT_ADD__}}
    self.Xclock.Add(self.Xport)
    // New SubPorts
{{__XPORT_CASCADED_SGN__}}
    return self
}

/*******************************/
/**      User APIs             */
/*******************************/
func (self *UT_{{__TOP_MODULE_NAME__}}) Step(a ...interface{}) {
    argc := len(a)
	if argc == 1 {
        self.Xclock.Step(a[0].(int))
    } else {
        self.Xclock.Step(1)
    }
}

func (self *UT_{{__TOP_MODULE_NAME__}}) Finish() {
    self.Dut.Finish()
}

func (self *UT_{{__TOP_MODULE_NAME__}}) InitClock(name string){
    self.Xclock.Add(self.Xport.Get(name))
}

func (self *UT_{{__TOP_MODULE_NAME__}}) StepRis(fc func(uint64)){
    self.Xclock.StepRis(fc)
}

func (self *UT_{{__TOP_MODULE_NAME__}}) StepFal(fc func(uint64)){
    self.Xclock.StepFal(fc)
}

func (self *UT_{{__TOP_MODULE_NAME__}}) OpenWaveform() bool {
    return self.Dut.OpenWaveform()
}

func (self *UT_{{__TOP_MODULE_NAME__}}) CloseWaveform() bool {
    return self.Dut.CloseWaveform()
}

func (self *UT_{{__TOP_MODULE_NAME__}}) GetXClock() xspcomm.XClock {
    return self.Xclock
}

func (self *UT_{{__TOP_MODULE_NAME__}}) GetXPort() xspcomm.XPort {
    return self.Xport
}

func (self *UT_{{__TOP_MODULE_NAME__}}) SetWaveform(filename string){
    self.Dut.SetWaveform(filename)
}

func (self *UT_{{__TOP_MODULE_NAME__}}) FlushWaveform(){
    self.Dut.FlushWaveform()
}

func (self *UT_{{__TOP_MODULE_NAME__}}) CheckPoint(name string) int {
    return self.Dut.CheckPoint(name)
}

func (self *UT_{{__TOP_MODULE_NAME__}}) Restore(name string) int {
    return self.Dut.Restore(name)
}

func (self *UT_{{__TOP_MODULE_NAME__}}) GetInternalSignal(name string, exargs ...interface{}) xspcomm.XData {
    // args: string name, int index = -1, bool use_vpi = false
    if v, ok := self.internalSignals[name]; ok {
        return v
    }
    var index int = -1
    var use_vpi bool = false
    var signal xspcomm.XData = nil
    var xname string = "CFG:" + name
    if len(exargs) > 0 {
        if len(exargs) == 1 {
            index = exargs[0].(int)
        } else if len(exargs) == 2 {
            index = exargs[0].(int)
            use_vpi = exargs[1].(bool)
        }
    }
    if !use_vpi {
        var signalGo xspcomm.XDataGo
        if index >= 0 {
            signalGo = self.Xcfg.NewXData(name, index, xname);
        } else {
            signalGo = self.Xcfg.NewXData(name, xname);
        }
        if signalGo == nil {
            return nil
        }
        signal = xspcomm.NewXDataFrom(signalGo)
    }else{
        signal = xspcomm.XDataFromVPI(self.Dut.GetVPIHandleObj(name),
                                      self.Dut.GetVPIFuncPtr("vpi_get"),
                                      self.Dut.GetVPIFuncPtr("vpi_get_value"),
                                      self.Dut.GetVPIFuncPtr("vpi_put_value"), "VPI:" + name)
    }
    if signal == nil {
        return nil
    }
    self.internalSignals[name] = signal
    return self.internalSignals[name]
}

func (self *UT_{{__TOP_MODULE_NAME__}}) GetInternalSignalArray(name string) *list.List {
    if v, ok := self.internalSignalsList[name]; ok {
        return v
    }
    return_list := list.New()
    signal_array := self.Xcfg.NewXDataArray(name, "CFG:" + name)
    if signal_array.Size() == 0 {
        return return_list
    }
    for i := 0; i < int(signal_array.Size()); i++ {
        return_list.PushBack(signal_array.Get(i))
    }
    self.internalSignalsList[name] = return_list
    return return_list
}

func (self *UT_{{__TOP_MODULE_NAME__}}) GetInternalSignalList(args ...interface{}) *list.List {
    var prefix string = ""
    var deep int = 99
    var use_vpi bool = false
    if len(args) > 0 {
        prefix = args[0].(string)
        if len(args) > 1 {
            deep = args[1].(int)
        }
        if len(args) > 2 {
            use_vpi = args[2].(bool)
        }
    }
    rlist := list.New()
    if !use_vpi {
        slist := self.Xcfg.GetSignalNames(prefix)
        if slist.Size() == 0 {
            return rlist
        }
        for i := 0; i < int(slist.Size()); i++ {
            signal := slist.Get(i)
            rlist.PushBack(signal)
        }
        return rlist
    }
    return self.VPIInternalSignalList(prefix, deep)
}

func (self *UT_{{__TOP_MODULE_NAME__}}) VPIInternalSignalList(args ...interface{}) *list.List {
    var prefix string = ""
    var deep int = 99
    if len(args) > 0 {
        prefix = args[0].(string)
        if len(args) > 1 {
            deep = args[1].(int)
        }
    }
    rlist := list.New()
    slist := self.Dut.VPIInternalSignalList(prefix, deep)
    if slist.Size() == 0 {
        return rlist
    }
    for i := 0; i < int(slist.Size()); i++ {
        rlist.PushBack(slist.Get(i))
    }
    return rlist
}

func (self *UT_{{__TOP_MODULE_NAME__}}) SetCoverage(filename string){
    self.Dut.SetCoverage(filename)
}

func (self *UT_{{__TOP_MODULE_NAME__}}) RefreshComb(){
    self.Dut.RefreshComb()
}

/*******************************/
/**    End of User APIs        */
/*******************************/

File: template/scala/dut.scala
package com.ut.{{__TOP_MODULE_NAME__}}


import com.xspcomm._

class UT_{{__TOP_MODULE_NAME__}}(args: Array[String]) extends JavaUT_{{__TOP_MODULE_NAME__}} with BaseDUTTrait {

    val vec = new StringScalaVector()
    for (i <- args.indices) {
      vec.add(args(i))
    }
    var dut = new DutUnifiedBase(vec)
    var xclock = new XClock(this.dut.getPxcStep(), this.dut.getPSelf())
    val internalSignals: scala.collection.mutable.Map[String, XData] = scala.collection.mutable.Map()
    val internalSignalsList: scala.collection.mutable.Map[String, Array[XData]] = scala.collection.mutable.Map()
    var xport = new XPort()
    var xcfg = this.newXCFG()

    def newXCFG(): XSignalCFG = {
        var ret: XSignalCFG = null
        try {
            JavaUT_{{__TOP_MODULE_NAME__}}.loadFileInJar("/{{__TOP_MODULE_NAME__}}_offset.yaml", (p) => {
                ret = new XSignalCFG(p, dut.GetXSignalCFGBasePtr())
            })
        } catch {
            case e: Exception => {
                if(dut.GetXSignalCFGBasePtr() == 0){
                    println("Error: " + e.getMessage)
                    return null
                }
                ret = new XSignalCFG("/{{__TOP_MODULE_NAME__}}_offset.yaml", dut.GetXSignalCFGBasePtr())
            }
        }
        ret
    }

    def this() = {
        this(Array.empty[String])
    }

    def this(arg: String) = {
        this(Array(arg))
    }

    // Create Pins
{{__XDATA_INIT__}}

    // Bind DPIs
{{__XDATA_BIND__}}

    // Add to port
{{__XPORT_ADD__}}

    // New Subports
{{__XPORT_CASCADED_SGN__}}

    this.xclock.Add(this.xport)
    
    /*************************************************/
    /*                  User APIs                    */
    /*************************************************/
    def GetXClock(): XClock = {
        this.xclock
    }
    def GetXPort(): XPort = {
        this.xport
    }
    def OpenWaveform(): Boolean = {
        this.dut.OpenWaveform()
    }
    def CloseWaveform(): Boolean = {
        this.dut.CloseWaveform()
    }
    def SetWaveform(wave_name: String) = {
        this.dut.SetWaveform(wave_name)
    }
    def FlushWaveform() = {
        this.dut.FlushWaveform()
    }
    def SetCoverage(coverage_name: String) = {
        this.dut.SetCoverage(coverage_name)
    }
    def Step(i: Int = 1): Unit = {
        this.xclock.Step(i)
    }
    def StepRis(callback: (Long) => Unit) = {
        this.xclock.StepRis(callback)
    }
    def StepFal(callback: (Long) => Unit) = {
        this.xclock.StepFal(callback)
    }
    def Finish(): Unit  = {
        this.dut.Finish()
    }
    def InitClock(clock_name: String) = {
        this.xclock.Add(this.xport.Get(clock_name))
    }
    def RefreshComb() = {
        this.dut.RefreshComb()
    }
    def CheckPoint(check_point: String) = {
        this.dut.CheckPoint(check_point)
    }
    def Restore(check_point: String) = {
        this.dut.Restore(check_point)
    }
    def VPIInternalSignalList(prefix: String="", deep: Int=99): Array[String] = {
        var ret: Array[String] = Array.empty[String]
        this.dut.VPIInternalSignalList(prefix, deep).forEach { i =>
            if (i != null) {
                ret :+= i
            }
        }
        ret
    }
    def GetInternalSignal(name: String, index: Int = -1, use_vpi: Boolean = false): XData = {
        if (this.internalSignals.contains(name)) {
            return this.internalSignals(name)
        }
        var signal: XData = null
        if(!use_vpi){
            val xname = "CFG:" + name
            if (dut.GetXSignalCFGBasePtr() == 0) {
                return signal
            }
            if (index >= 0) {
                signal = this.xcfg.NewXData(name, index, xname)
            }else{
                signal = this.xcfg.NewXData(name, xname)
            }
            if (signal != null) {
                this.internalSignals += (name -> signal)
            }
            return signal
        }
        signal = XData.FromVPI(this.dut.GetVPIHandleObj(name),
                                   this.dut.GetVPIFuncPtr("vpi_get"),
                                   this.dut.GetVPIFuncPtr("vpi_get_value"),
                                   this.dut.GetVPIFuncPtr("vpi_put_value"), "VPI:" + name)
        if (signal != null) {
            this.internalSignals += (name -> signal)
        }
        signal
    }
    def GetInternalSignal(name: String, is_array:Boolean): Array[XData] = {
        if (this.internalSignalsList.contains(name)) {
            return this.internalSignalsList(name)
        }
        var result = Array.empty[XData]
        if (this.dut.GetXSignalCFGBasePtr() == 0) {
            return result
        }
        if(!is_array){
            return result
        }
        val xname = "CFG:" + name
        val signalList = this.xcfg.NewXDataArray(name, xname)
        if (signalList != null) {
            signalList.forEach { i =>
                if (i != null) {
                    result :+= i
                }
            }
        }
        result
    }
    def GetInternalSignalList(prefix:String="", deep:Int=99, use_vpi:Boolean=false): Array[String] = {
        var ret: Array[String] = Array.empty[String]
        if (this.dut.GetXSignalCFGBasePtr() != 0 && !use_vpi) {
            this.xcfg.GetSignalNames(prefix).forEach { i =>
                if (i != null) {
                    if (prefix == "" || i.startsWith(prefix)) {
                        ret :+= i
                    }
                }
            }
            return ret
        }
        this.VPIInternalSignalList(prefix, deep)
    }
    /*************************************************/
    /*               End of User APIs                */
    /*************************************************/
}
